File: Concurrency/1195. Fizz Buzz Multithreaded.py
import threading
class FizzBuzz:
    def __init__(self, n: int):
        self.n = n
        self.cv = threading.Condition()
        self.count = 1

    # printFizz() outputs "fizz"
    def fizz(self, printFizz: 'Callable[[], None]') -> None:
        while True:
            with self.cv:
                while (self.count % 3 != 0 or self.count % 5 ==0)  and self.count <= self.n:
                    self.cv.wait()
                if self.count > self.n:
                    return
                printFizz()
                self.count +=  1
                self.cv.notifyAll()

    # printBuzz() outputs "buzz"
    def buzz(self, printBuzz: 'Callable[[], None]') -> None:
        while True:
            with self.cv:
                while (self.count % 3 == 0 or self.count % 5 !=0)  and self.count <= self.n:
                    self.cv.wait()
                if self.count > self.n:
                    return
                printBuzz()
                self.count +=  1
                self.cv.notifyAll()

    # printFizzBuzz() outputs "fizzbuzz"
    def fizzbuzz(self, printFizzBuzz: 'Callable[[], None]') -> None:
        while True:
            with self.cv:
                while (self.count % 3 != 0 or self.count % 5 !=0 ) and self.count <= self.n:
                    self.cv.wait()
                if self.count > self.n:
                    return
                printFizzBuzz()
                self.count +=  1
                self.cv.notifyAll()

    # printNumber(x) outputs "x", where x is an integer.
    def number(self, printNumber: 'Callable[[int], None]') -> None:
        while True:
            with self.cv:
                while (self.count % 3 == 0 or self.count % 5 ==0)  and self.count <= self.n:
                    self.cv.wait()
                if self.count > self.n:
                    return
                printNumber(self.count)
                self.count +=  1
                self.cv.notifyAll()



import threading

class FizzBuzz:
    def __init__(self, n):
        self.n = n
        self.f = threading.Lock()
        self.b = threading.Lock()
        self.fb = threading.Lock()
        self.main = threading.Lock()        
        self.f.acquire()
        self.b.acquire()
        self.fb.acquire()

    # printFizz() outputs "fizz"
    def fizz(self, printFizz):
        for i in range (self.n // 3 - self.n // 15):
            self.f.acquire()
            printFizz()
            self.main.release()
            
    # printBuzz() outputs "buzz"
    def buzz(self, printBuzz):
        for i in range (self.n // 5 - self.n // 15):
            self.b.acquire()
            printBuzz()
            self.main.release()

    # printFizzBuzz() outputs "fizzbuzz"
    def fizzbuzz(self, printFizzBuzz):
        for i in range (self.n // 15):
            self.fb.acquire()
            printFizzBuzz()
            self.main.release()        

    # printNumber(x) outputs "x", where x is an integer.
    def number(self, printNumber):
        for i in range(1, self.n + 1):         
            self.main.acquire()        
            if i % 15 == 0:
                self.fb.release()
            elif i % 3 == 0:
                self.f.release()
            elif i % 5 == 0:
                self.b.release()
            else:
                printNumber(i)
                self.main.release()   



## Semaphore
class FizzBuzz:
    def __init__(self, n: int):
        self.n = n
        self.semaphore_0 = threading.Semaphore(1)
        self.semaphore_3 = threading.Semaphore(0)
        self.semaphore_5 = threading.Semaphore(0)
        self.semaphore_15 = threading.Semaphore(0)

    # printFizz() outputs "fizz"
    def fizz(self, printFizz: 'Callable[[], None]') -> None:
    	for i in range(3, self.n+ 1, 3):
            if i % 15:
                self.semaphore_3.acquire()
                printFizz()
                self.release_semaphore(i+1)

    # printBuzz() outputs "buzz"
    def buzz(self, printBuzz: 'Callable[[], None]') -> None:
    	for i in range(5, self.n+ 1, 5):
            if i % 15:
                self.semaphore_5.acquire()
                printBuzz()
                self.release_semaphore(i+1)

    # printFizzBuzz() outputs "fizzbuzz"
    def fizzbuzz(self, printFizzBuzz: 'Callable[[], None]') -> None:
        for i in range(15, self.n+ 1, 15):
            self.semaphore_15.acquire()
            printFizzBuzz()
            self.release_semaphore(i+1)

    # printNumber(x) outputs "x", where x is an integer.
    def number(self, printNumber: 'Callable[[int], None]') -> None:
        for i in range(1, self.n + 1):
            if i % 3 and i % 5:
                self.semaphore_0.acquire()
                printNumber(i)
                self.release_semaphore(i+1)

    
    def release_semaphore(self, i):
        if i %3 == 0 and i % 5:
            self.semaphore_3.release()
        elif i % 5 == 0 and i % 3:
            self.semaphore_5.release() 
        elif i%15 == 0:
            self.semaphore_15.release()
        else:
            self.semaphore_0.release()




import threading

class FizzBuzz:
    def __init__(self, n: int):      
        self.n = n
        self.f = threading.Lock()
        self.b = threading.Lock()
        self.fb = threading.Lock()
        self.f.acquire()
        self.b.acquire()
        self.fb.acquire()
        self.main = threading.Lock()

    # printFizz() outputs "fizz"
    def fizz(self, printFizz: 'Callable[[], None]') -> None:
        while True:
            self.f.acquire()
            if self.n == 0 :
                return
            printFizz()
            self.main.release()

    # printBuzz() outputs "buzz"
    def buzz(self, printBuzz: 'Callable[[], None]') -> None:
        while True:
            self.b.acquire()
            if self.n == 0:
                return
            printBuzz()
            self.main.release()

    # printFizzBuzz() outputs "fizzbuzz"
    def fizzbuzz(self, printFizzBuzz: 'Callable[[], None]') -> None:
        while True:
            self.fb.acquire()
            if self.n == 0:
                return
            printFizzBuzz()
            self.main.release()

    # printNumber(x) outputs "x", where x is an integer.
    def number(self, printNumber: 'Callable[[int], None]') -> None:
        for i in range(1, self.n+1):
            #print("number:",i)
            self.main.acquire()
            if i % 15 == 0:
                self.fb.release()
            elif i % 3 == 0:
                self.f.release()
            elif i % 5 == 0:
                self.b.release()
            else:
                printNumber(i)
                self.main.release()

        self.main.acquire() 
        self.n = 0
        self.f.release()
        self.b.release()
        self.fb.release()
        return

File: Concurrency/1195. Fizz Buzz Multithreaded.cpp
class FizzBuzz {
private:
    int n;
    int i=1;
    condition_variable cv;
    mutex mu;

public:
    FizzBuzz(int n) {
        this->n = n;
    }
    
    void run(function<bool(int)>test, function<void(int)>print){
        while(true){
            unique_lock<mutex>lock(mu);
            cv.wait(lock, [&](){
               return  i > n || test(i);
            });
            if(i > n)
                break;
            print(i++);
            cv.notify_all();
        }
        
    }

    // printFizz() outputs "fizz".
    void fizz(function<void()> printFizz) {
        run([](int i){return i%3==0 && i%5!=0;},[&](int i){ printFizz();});
    }

    // printBuzz() outputs "buzz".
    void buzz(function<void()> printBuzz) {
        run([](int i){return i%3!=0 && i%5 == 0;}, [&](int i){printBuzz();});
    }

    // printFizzBuzz() outputs "fizzbuzz".
	void fizzbuzz(function<void()> printFizzBuzz) {
        run([](int i){return i%3==0 && i%5 == 0;}, [&](int i){printFizzBuzz();});
    }

    // printNumber(x) outputs "x", where x is an integer.
    void number(function<void(int)> printNumber) {
        run([](int i){return i%3!=0 && i%5 != 0;}, printNumber);
    }
};



class FizzBuzz {
private:
    int n;
    int count;
    mutex m;
    condition_variable cv;

public:
    FizzBuzz(int n) {
        this->n = n;
        this->count = 1;
    }

    void fizz(function<void()> printFizz) {
        while (true) {
            unique_lock<mutex> lock(m);
            cv.wait(lock,[&](){
                return count > n || (count % 3 == 0 && count % 5 != 0);
            });
            if (count > n) return;
            printFizz();
            ++count;
            cv.notify_all();
        }
    }

    void buzz(function<void()> printBuzz) {
        while (true) {
            unique_lock<mutex> lock(m);
            cv.wait(lock,[&](){
                return count > n || (count % 3 != 0 && count % 5 == 0);
            });
            if (count > n) return;
            printBuzz();
            ++count;
            cv.notify_all();
        }
    }

	void fizzbuzz(function<void()> printFizzBuzz) {
        while (true) {
            unique_lock<mutex> lock(m);
            cv.wait(lock,[&](){
                return count > n || (count % 5 == 0 && count % 3 == 0);
            });
            if (count > n) return;
            printFizzBuzz();
            ++count;
            cv.notify_all();
        }
    }

    void number(function<void(int)> printNumber) {
        while (true) {
            unique_lock<mutex> lock(m);
            cv.wait(lock,[&](){
                return count > n || (count % 5 != 0 && count % 3 != 0);
            });
            if (count > n) return;
            printNumber(count++);
            cv.notify_all();
        }
    }
};

/*

 //可以更改四个thread 的顺序
比如

 printFizz cv 条件 curr_ % 4 == 1；
 printBuzz cv 条件 curr_ % 4 == 0；
 printFizzBuzz cv 条件 curr % 4 = 3
 printNumber  cv 条件 curr % 4 = 2

 四个thread的wait 让每个i都必须走这四个function一次, 从而按照顺序打印

 比如 
 curr = 0, 先到 printBuzz， i = 1, 不打印 
 curr = 1, 到 printFizz,   i = 1 不打印
 curr = 2, 到 printNumber,  i = 1, 打印
 curr = 3, 到printFizzBuzz， i = 1 不打印

curr = 4, 先到 printBuzz， i = 2, 不打印 
curr = 5, 到 printFizz,   i = 2 不打印
curr = 6, 到 printNumber,  i = 2, 打印
curr = 7, 到printFizzBuzz， i = 2 不打印


curr = 8, 先到 printBuzz， i = 3, 打印
curr = 9, 到 printFizz,   i = 3, 不打印
curr = 10, 到 printNumber,  i = 3, 不打印
curr = 11, 到printFizzBuzz， i = 3 不打印
*/


class FizzBuzz {
public:
    FizzBuzz(int n) : n_(n) {
    }

    // printFizz() outputs "fizz".
    void fizz(function<void()> printFizz) {
        for (int i = 1; i <= n_; ++i) {
            {
                unique_lock<mutex> l(m_);
                cv_.wait(l, [this]() { return curr_ % 4 == 0; });
                ++curr_;
                if (i % 3 == 0 && i % 5 != 0) {
                    printFizz();
                }
            }
            cv_.notify_all();
        }
    }

    // printBuzz() outputs "buzz".
    void buzz(function<void()> printBuzz) {
        for (int i = 1; i <= n_; ++i) {
            {
                unique_lock<mutex> l(m_);
                cv_.wait(l, [this]() { return curr_ % 4 == 1; });
                ++curr_;
                if (i % 3 != 0 && i % 5 == 0) {
                    printBuzz();
                }
            }
            cv_.notify_all();
        }
    }

    // printFizzBuzz() outputs "fizzbuzz".
    void fizzbuzz(function<void()> printFizzBuzz) {
        for (int i = 1; i <= n_; ++i) {
            {
                unique_lock<mutex> l(m_);
                cv_.wait(l, [this]() { return curr_ % 4 == 2; });
                ++curr_;
                if (i % 3 == 0 && i % 5 == 0) {
                    printFizzBuzz();
                }
            }
            cv_.notify_all();
        }
    }

    // printNumber(x) outputs "x", where x is an integer.
    void number(function<void(int)> printNumber) {
        for (int i = 1; i <= n_; ++i) {
            {
                unique_lock<mutex> l(m_);
                cv_.wait(l, [this]() { return curr_ % 4 == 3; });
                ++curr_;
                if (i % 3 != 0 && i % 5 != 0) {
                    printNumber(i);
                }
            }
            cv_.notify_all();
        }
    }

private:
    int n_;
    int curr_ = 0;
    mutex m_;
    condition_variable cv_;
};
